### PR TITLE
istioctl experimental metrics: add service metrics support

### PR DESCRIPTION
**Pull request by nicole-lihui**
**Please provide a description of this PR:**
user-facing: 
  ```
  # istioctl experimental metrics support format:  [<resource type>/]<name>[.<namespace>]


  # Retrieve workload metrics for productpage-v1 workload with resource type [workload] 
  istioctl experimental metrics workload/productpage-v1

  # Retrieve service metrics for productpage with resource type [service] 
  istioctl experimental metrics service/productpage

  # Retrieve service metrics for various services in the different namespaces
  istioctl experimental metrics service/productpage.foo

  # Retrieve workload metrics for various services with resource type [service] in the different namespaces
  istioctl experimental metrics service/productpage.foo service/reviews.bar
  ```

more info:
issue: https://github.com/istio/istio/issues/36428
doc:  https://docs.google.com/document/d/1ATp4qCHktncfAFjH_S9yWm9dGxI_aXHN2cw1AD4kveQ/edit?usp=sharing

### CODE REVIEW

**Comment by nicole-lihui** [comment]
@zirain 
PR https://github.com/istio/istio/pull/37822  not stale
Before the  PR has a problem, submit a new PR, If you have interest can follow up this PR, Thank you 😄 。

File: istioctl/cmd/metics-query.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
+// Copyright Istio Authors
```
Comment (zirain) [inline istioctl/cmd/metics-query.go:1]: nit: filename should be `metric_query.go`

File: istioctl/cmd/metrics.go [hunk 2]
```diff
@@ -19,15 +19,18 @@ import (
 	"errors"
 	"fmt"
 	"io"
-	"strings"
 	"text/tabwriter"
 	"time"
 
+	"istio.io/istio/istioctl/pkg/util/handlers"
```
Comment (zirain) [inline istioctl/cmd/metrics.go:25]: nit: sort pkg

File: tests/integration/telemetry/stats/prometheus/nullvm/istioctl_metrics_test.go [hunk 3]
```diff
@@ -56,7 +56,7 @@ func validateDefaultOutput(t framework.TestContext, workload string) error { //
 	}
 
 	// output will be something like
-	//      WORKLOAD    TOTAL RPS    ERROR RPS  P50 LATENCY  P90 LATENCY  P99 LATENCY
+	//          NAME    TOTAL RPS    ERROR RPS  P50 LATENCY  P90 LATENCY  P99 LATENCY
```
Comment (zirain) [inline tests/integration/telemetry/stats/prometheus/nullvm/istioctl_metrics_test.go:59]: do we need an new column like `resource type` or something?

File: istioctl/pkg/util/handlers/handlers.go [hunk 4]
```diff
@@ -62,6 +62,22 @@ func inferNsInfo(name, namespace string) (string, string) {
 	return name[0:separator], name[separator+1:]
 }
 
+func InferResourceTuple(name, specifyNS string) (resType string, resName string, resNamespace string, err error) {
```
Comment (zirain) [inline istioctl/pkg/util/handlers/handlers.go:65]: nit: do no export this function

File: istioctl/cmd/metics-query.go [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+type Labels []string
+
+type LabelsBuilder struct {
+	LabelSelector labels.Selector
+	ByLabels      Labels
+	Dur           string
```
Comment (zirain) [inline istioctl/cmd/metics-query.go:28]: what's is this short for?

File: istioctl/cmd/metics-query.go [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+import (
+	"fmt"
+
+	"github.com/prometheus/prometheus/model/labels"
+)
+
+type Labels []string
+
+type LabelsBuilder struct {
```
Comment (zirain) [inline istioctl/cmd/metics-query.go:25]: do prometheus have same struct for this?

File: istioctl/cmd/metics-query.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+		if i < len(ls)-1 {
+			str += l + ","
+		} else {
+			str += l
+		}
+	}
+	return str
+}
+
+type p8sQuantile float64
```
Comment (zirain) [inline istioctl/cmd/metics-query.go:113]: why do we need this? any special reason?

File: istioctl/cmd/metics-query.go (hunk 6 above)
Comment (nicole-lihui) [reply istioctl/cmd/metics-query.go:25]: Promethus do not have the same struct, this structure named possible ambiguity.

like QueryVariableBuilder（Temporarily didn't think of a more appropriate name）

Responsible for the structural query variable， example : sum(rate(istio_request_total{%s}[%s])) by (%s)
Doing so will help keep the query more flexible。

File: istioctl/cmd/metics-query.go (hunk 5 above)
Comment (nicole-lihui) [reply istioctl/cmd/metics-query.go:28]: duration
this should be use long name

File: istioctl/cmd/metics-query.go (hunk 7 above)
Comment (nicole-lihui) [reply istioctl/cmd/metics-query.go:113]: Personal habits will have a `hard code` associated with structure definition, prevent wrong input

File: tests/integration/telemetry/stats/prometheus/nullvm/istioctl_metrics_test.go (hunk 3 above)
Comment (nicole-lihui) [reply tests/integration/telemetry/stats/prometheus/nullvm/istioctl_metrics_test.go:59]: I reference `kubectl get` cmd，only input/output  `Name` will be more intuitive.

```bash
kubectl get service/details  deployment.apps/details-v1
NAME              TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/details   ClusterIP   10.104.202.24   <none>        9080/TCP   30d

NAME                         READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/details-v1   1/1     1            1           30d
```

Comment (zirain) [reply tests/integration/telemetry/stats/prometheus/nullvm/istioctl_metrics_test.go:59]: agree.

**Comment by nicole-lihui** [comment]
@douglas-reid The PR in line with expectations?

**Comment by douglas-reid** [comment]
Sorry, I missed this activity. Still interested in merging?

**Comment by nicole-lihui** [comment]
> Sorry, I missed this activity. Still interested in merging?

@douglas-reid  it doesn't matter ~

I am very interested in it

**Comment by obaranov1** [comment]
@nicole-lihui, @douglas-reid, @zirain  Hello! What happened to this PR and why it was not merged?

**Pull request closed without merging**